test/powershell/engine/Api/TaskBasedAsyncPowerShellAPI.Tests.ps1

```diff
@@ -0,0 +1,251 @@
+            } finally {
+                $ps.Dispose()
+            }
+        }
+
+        It 'handles terminating errors properly when invoked asynchronously in a new runspace' {
+            $ps = [powershell]::Create()
+            try {
+                $sb = {
+                    $r = $ps.AddScript(@'
+try {
+    Get-Process -Invalid 42 -ErrorAction Stop
+} catch {
+    throw
+}
+'@).InvokeAsync()
+                    Set-Variable -Name r -Scope 2 -Value $r
+                    [System.Threading.Tasks.Task]::WaitAll(@($r))
+                    $r
+                }
+                # This test is designed to gracefully fail with an error when invoked asynchronously.
+                { $sb.Invoke() } | Should -Throw -ErrorId 'AggregateException'
+                $r.IsFaulted | Should -Be $true
+                $r.Exception.InnerException -is [System.Management.Automation.ParameterBindingException] | Should -Be $true
+                $r.Exception.InnerException.CommandInvocation.InvocationName | Should -BeExactly 'Get-Process'
+                $r.Exception.InnerException.ParameterName | Should -BeExactly 'Invalid'
+                $r.Exception.InnerException.ErrorId | Should -BeExactly 'NamedParameterNotFound'
+            } finally {
+                $ps.Dispose()
+            }
+        }
+
+        It 'cannot invoke a single script asynchronously in a runspace that has not been opened' {
+            $rs = [runspacefactory]::CreateRunspace()
+            $ps = [powershell]::Create($rs)
+            try {
+                $r = $ps.AddScript('@(1..10).foreach{Start-Sleep -Milliseconds 500}').InvokeAsync()
+                # This test is designed to fail. You cannot invoke PowerShell asynchronously
+                # in a runspace that has not been opened.
+                $r.IsFaulted | Should -Be $true
+                $r.Exception -is [System.AggregateException] | Should -Be $true
+                $r.Exception.InnerException -is [System.Management.Automation.Runspaces.InvalidRunspaceStateException] | Should -Be $true
+                $r.Exception.InnerException.CurrentState | Should -Be 'BeforeOpen'
+                $r.Exception.InnerException.ExpectedState | Should -Be 'Opened'
+            } finally {
+                $ps.Dispose()
+                $rs.Dispose()
+            }
+        }
+
+        It 'cannot invoke a single script asynchronously in a runspace that is busy' {
+            $rs = [runspacefactory]::CreateRunspace()
+            $rs.Open()
+            $psBusy = [powershell]::Create($rs)
+            try {
+                $psBusy.AddScript('@(1..120).foreach{Start-Sleep -Milliseconds 500}').InvokeAsync() > $null
+                $time = 0
+                while ($rs.RunspaceAvailability -ne 'Busy') {
+                    if (($time += 100) -ge 120000) {
+                        break
+                    }
+                    Start-Sleep -Milliseconds 100
+                }
```

Comment:
I have that in the next text (cannot invoke a single script asynchronously in the current runspace). I thought there was some redundancy here as well, however since @BrucePay [requested tests with the current runspace or another runspace that is busy](https://github.com/PowerShell/PowerShell/pull/8056#discussion_r247385023), I put both tests in place to ensure his concerns were covered.

Reply:
Using the current Runspace also satisfies Bruce's ask.
```
$ps = [powershell]::Create()
$ps.Runspace = $Host.Runspace
$ps.AddScript("1+1").Invoke()

> "The pipeline was not run because a pipeline is already running. Pipelines cannot be run concurrently."
```

Reply:
Fair enough, change made.